README.md
`toy-tax-calculator`
====================

Calculates fictitious taxes given gross income and a tax year.

Usage
-----

```
$ toy-tax-calculator --help
Calculates (fictitious) taxes based on gross income for a given year

Usage:
  toy-tax-calculator [flags]

Flags:
      --gross-income string   Total amount of gross income (default "0")
  -h, --help                  help for toy-tax-calculator
      --tax-year int          Year where taxes apply (default 2019)
```

Developing
----------

### Requirements

* Go ≥ 1.13
* _(optionally)_ Podman or Docker
* GNU Make

### Building

```
make
```

Optionally if using Podman or Docker:

```
podman build -t toy-tax-calculator:develop .
```

Either build runs all tests before generating binary artifacts.

TODOs
-----

* CI/CD public pipelines
* Release binaries


CHANGELOG.md
<a name="1.0.0"></a>
## 1.0.0 (2019-09-09)


#### Features

*   add tax years for 2018, 2017 and 2016 ([9922b1d1](9922b1d1))
*   add tests to `Makefile` target and always run it ([68374054](68374054))
*   calculate taxes for year 2019 ([f25fb2a9](f25fb2a9))
*   make tax dues printable, but without implementation ([5a61d05c](5a61d05c))
*   add tax structure for 2019 ([b5b41924](b5b41924))
*   add warning that calculating taxes is not done ([3dccea6d](3dccea6d))
*   add support for tool version pinning using `asdf` ([5bdecd35](5bdecd35))
*   define types for tax calculation ([00051873](00051873))
*   treat income as string during CLI input ([e7f1a88e](e7f1a88e))
*   add specification under `docs` ([3f7251c8](3f7251c8))
*   print usage from `--help` flag ([e9d516c1](e9d516c1))





cmd/tax_test.go
package cmd

import (
	"testing"

	taxInternal "github.com/kerscher/toy-tax-calculator/internal/app"
	"github.com/kerscher/toy-tax-calculator/pkg/tax"
	"github.com/shopspring/decimal"
)

func TestTaxMatchesExample(t *testing.T) {
	inGrossIncome := tax.GrossIncome{
		Value:    decimal.NewFromFloat(43_500),
		Currency: taxInternal.DefaultCurrency,
	}
	inYear := 2019
	if inTaxYear, ok := taxInternal.Rates[inYear]; ok {
		want := tax.PayerYear{
			Year:        2019,
			Allowance:   inTaxYear.Allowance,
			GrossIncome: inGrossIncome,
			TaxableIncome: tax.TaxableIncome{
				Value:    decimal.NewFromFloat(31_650),
				Currency: taxInternal.DefaultCurrency,
			},
			Due: tax.Amount{
				Value:    decimal.NewFromFloat(6_518.69),
				Currency: taxInternal.DefaultCurrency,
			},
			PayerBands: []tax.PayerBand{
				tax.PayerBand{
					Band: tax.Band{
						Description: "Starter rate",
						Rate:        tax.Rate(decimal.NewFromFloat(0.19)),
						Capacity: tax.Amount{
							Value:    decimal.NewFromFloat(2_000),
							Currency: taxInternal.DefaultCurrency,
						},
					},
					Due: tax.Amount{
						Value:    decimal.NewFromFloat(380),
						Currency: taxInternal.DefaultCurrency,
					},
				},
				tax.PayerBand{
					Band: tax.Band{
						Description: "Basic rate",
						Rate:        tax.Rate(decimal.NewFromFloat(0.20)),
						Capacity: tax.Amount{
							Value:    decimal.NewFromFloat(10_149),
							Currency: taxInternal.DefaultCurrency,
						},
					},
					Due: tax.Amount{
						Value:    decimal.NewFromFloat(2_029.80),
						Currency: taxInternal.DefaultCurrency,
					},
				},
				tax.PayerBand{
					Band: tax.Band{
						Description: "Intermediate rate",
						Rate:        tax.Rate(decimal.NewFromFloat(0.21)),
						Capacity: tax.Amount{
							Value:    decimal.NewFromFloat(19_429),
							Currency: taxInternal.DefaultCurrency,
						},
					},
					Due: tax.Amount{
						Value:    decimal.NewFromFloat(4_080.09),
						Currency: taxInternal.DefaultCurrency,
					},
				},
				tax.PayerBand{
					Band: tax.Band{
						Description: "Higher rate",
						Rate:        tax.Rate(decimal.NewFromFloat(0.40)),
						Capacity: tax.Amount{
							Value:    decimal.NewFromFloat(118_419),
							Currency: taxInternal.DefaultCurrency,
						},
					},
					Due: tax.Amount{
						Value:    decimal.NewFromFloat(28.80),
						Currency: taxInternal.DefaultCurrency,
					},
				},
				tax.PayerBand{
					Band: tax.Band{
						Description: "Top rate",
						Rate:        tax.Rate(decimal.NewFromFloat(0.46)),
						Capacity:    taxInternal.MaxIncome,
					},
					Due: tax.Amount{
						Value:    decimal.NewFromFloat(0),
						Currency: taxInternal.DefaultCurrency,
					},
				},
			},
		}
		py, _ := tax.NewPayerYear(inTaxYear, inGrossIncome)
		// The condition below should use reflect.DeepEqual, but
		// despite having identical values DeepEqual is unable to
		// maintain equality. This is likely due to limitations with
		// how it handles pointer dereferencing for values inside
		// decimal.Decimal. To circumvent this we use the String
		// method for both.
		if py.String() != want.String() {
			t.Errorf("Calculated tax does not match reference.\n\nReference:\n----------\n%v\n\nCalculated:\n-----------\n %v", py, want)
			return
		}
		return
	}
	t.Errorf("Year %v rates does not exist for integration tests. Please rectify the internal library code.", inYear)
}


pkg/tax/tax_test.go
package tax

import (
	"testing"

	"github.com/leekchan/accounting"
	"github.com/shopspring/decimal"
)

var (
	testBand = Band{
		Description: "Testing rate",
		Rate:        Rate(decimal.NewFromFloat(0.5)),
		Capacity: Amount{
			Value:    decimal.NewFromFloat(1_000_000),
			Currency: accounting.DefaultAccounting("£", 2),
		},
	}
)

func TestDueIsZeroOnZeroAmount(t *testing.T) {
	in := Amount{
		Value:    decimal.NewFromFloat(0),
		Currency: testBand.Capacity.Currency,
	}
	pb, _, _ := testBand.due(in)
	if !decimal.Decimal(pb.Due.Value).Equal(decimal.NewFromFloat(0)) {
		t.Errorf("Zero income should lead to zero taxes due. Due: %v", pb.Due.Value)
	}
}

func TestDueIsCompatibleWithRate(t *testing.T) {
	in := Amount{
		Value:    decimal.NewFromFloat(500_000),
		Currency: testBand.Capacity.Currency,
	}
	want := decimal.NewFromFloat(250_000)
	pb, _, _ := testBand.due(in)
	if !decimal.Decimal(pb.Due.Value).Equal(want) {
		t.Errorf("Taxes due != (Income * Rate). Due: %v, Income: %v, Rate: %v, Expected: %v", pb.Due.Value, in, decimal.Decimal(pb.Rate).String(), want)
	}
}


pkg/tax/rate_test.go
package tax

import (
	"testing"

	"github.com/shopspring/decimal"
)

func TestRatesAreValidOrError(t *testing.T) {
	ts := []struct {
		in   decimal.Decimal
		want Rate
	}{
		{
			in:   decimal.NewFromFloat(0),
			want: Rate(decimal.NewFromFloat(0)),
		},
		{
			in:   decimal.NewFromFloat(0.5),
			want: Rate(decimal.NewFromFloat(0.5)),
		},
		{
			in:   decimal.NewFromFloat(32),
			want: Rate{},
		},
		{
			in:   decimal.NewFromFloat(-2389),
			want: Rate{},
		},
	}
	for _, c := range ts {
		got, _ := NewRate(c.in)
		if !decimal.Decimal(got).Equal(decimal.Decimal(c.want)) {
			t.Errorf("NewRate(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}


cmd/cmd.go
package cmd

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
)

// Cfg is a struct with configuration flags and arguments
type Cfg struct {
	taxYear     int
	grossIncome string
}

var (
	cfg Cfg
)

var rootCmd = &cobra.Command{
	Use:   "toy-tax-calculator",
	Short: "Calculates (fictitious) taxes",
	Long:  `Calculates (fictitious) taxes based on gross income for a given year`,
	Run:   calculateTax,
}

// Execute runs once and is called by cobra for CLI processing
func Execute() {
	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

func init() {
	rootCmd.PersistentFlags().IntVar(&cfg.taxYear, "tax-year", 2019, "Year where taxes apply")
	rootCmd.PersistentFlags().StringVar(&cfg.grossIncome, "gross-income", "0", "Total amount of gross income")
}

func processArgs(cmd *cobra.Command, args []string) (*Cfg, error) {
	y, err := cmd.Flags().GetInt("tax-year")
	if err != nil {
		return &Cfg{}, err
	}

	g, err := cmd.Flags().GetString("gross-income")
	if err != nil {
		return &Cfg{}, err
	}

	return &Cfg{taxYear: y, grossIncome: g}, nil
}


pkg/tax/tax.go
// Package tax provides currency-independent progressive tax
// functionality
//
// Internally everything is kept as decimal.Decimal, but users are
// required to also choose a currency when initialising and using
// values.
package tax

import (
	"fmt"
	"strings"

	"github.com/leekchan/accounting"
	"github.com/shopspring/decimal"
)

// Amount is a value in a given currency
type Amount struct {
	Value    decimal.Decimal
	Currency *accounting.Accounting
}

func (a Amount) String() string {
	return a.Currency.FormatMoneyDecimal(a.Value)
}

// Allowance is the maximum tax-exempt income a taxpayer has
type Allowance Amount

// TaxBand consists of a description, a rate and a capacity. The
// actual band is only meaningful in the context of other TaxBands, as
// a sequence of them naturally leads to a minimum and a maximum for a
// particular band depending on its capacity.
//
// This particular design choice ensures bands are never overlapping
// and ensures correctness by construction.
type Band struct {
	Description string
	Rate        Rate
	Capacity    Amount
}

// A PayerBand is a Band enriched with the amount due by a given taxpayer
type PayerBand struct {
	Band
	Due Amount
}

// Converts a Band into a PayerBand. This is unexported as the Amount
// passed to it must be enforced elsewhere, as nothing guarantees a
// user will keep invariants related to GrossIncome. The Amount
// returned is the residual Amount if it exceeds Capacity for that
// band.
func (b *Band) due(a Amount) (PayerBand, Amount, error) {
	if b.Capacity.Currency != a.Currency {
		return PayerBand{}, Amount{}, fmt.Errorf("Cannot calculate taxes due. Currencies for Band and income Amount are not identical.")
	}
	if a.Value.LessThanOrEqual(b.Capacity.Value) {
		return PayerBand{
				Band: *b,
				Due: Amount{
					Value:    a.Value.Mul(decimal.Decimal(b.Rate)),
					Currency: a.Currency,
				},
			}, Amount{
				Value:    decimal.NewFromFloat(0),
				Currency: a.Currency,
			}, nil
	}
	return PayerBand{
			Band: *b,
			Due: Amount{
				Value:    b.Capacity.Value.Mul(decimal.Decimal(b.Rate)),
				Currency: a.Currency,
			},
		}, Amount{
			Value:    a.Value.Sub(b.Capacity.Value),
			Currency: a.Currency,
		}, nil
}

func (pb PayerBand) String() string {
	return fmt.Sprintf("%v: %v", pb.Band.Description, Amount(pb.Due).String())
}

// A Year has an allowance value and an ordered sequence of TaxBand
type Year struct {
	Year      int
	Allowance Allowance
	Bands     []Band
}

// GrossIncome is the total income of a taxpayer before any taxes are
// calculated
type GrossIncome Amount

// TaxableIncome is the part of a taxpayer income due taxes in a given
// year
type TaxableIncome Amount

// Taxable converts GrossIncomes into TaxableIncomes
func (g *GrossIncome) Taxable(a Allowance) (TaxableIncome, error) {
	if g.Currency != a.Currency {
		return TaxableIncome{}, fmt.Errorf("Cannot calculate taxable income. Currencies for GrossIncome and Allowance are not identical.")
	}
	t := g.Value.Sub(a.Value)
	if t.IsNegative() {
		return TaxableIncome{
			Value:    decimal.NewFromFloat(0),
			Currency: g.Currency,
		}, nil
	}
	return TaxableIncome{
		Value:    t,
		Currency: g.Currency,
	}, nil
}

//

// PayerYear is a Year with a sequence of PayerBand instead of Band
// and payer income information
type PayerYear struct {
	Year          int
	Allowance     Allowance
	GrossIncome   GrossIncome
	TaxableIncome TaxableIncome
	Due           Amount
	PayerBands    []PayerBand
}

// Calculates taxes of a taxpayer given their gross income and a tax
// Year
func NewPayerYear(y Year, g GrossIncome) (PayerYear, error) {
	ti, err := g.Taxable(y.Allowance)
	if err != nil {
		return PayerYear{}, err
	}
	tti := ti
	d := Amount{Value: decimal.NewFromFloat(0), Currency: g.Currency}
	var pbs []PayerBand
	for _, b := range y.Bands {
		pb, _, err := b.due(Amount(tti))
		if err != nil {
			return PayerYear{}, err
		}
		ntv := tti.Value.Sub(b.Capacity.Value)
		if ntv.IsNegative() {
			ntv = decimal.NewFromFloat(0)
		}
		tti = TaxableIncome{
			Value:    ntv,
			Currency: tti.Currency,
		}
		d.Value = d.Value.Add(pb.Due.Value)
		pbs = append(pbs, pb)
	}
	return PayerYear{
		Year:          y.Year,
		Allowance:     y.Allowance,
		GrossIncome:   g,
		TaxableIncome: ti,
		Due:           d,
		PayerBands:    pbs,
	}, nil
}

func (p PayerYear) String() string {
	var pbstr strings.Builder
	for _, pb := range p.PayerBands {
		pbstr.WriteString(pb.String())
		pbstr.WriteString("\n")
	}
	return fmt.Sprintf(`
Tax year: %d

Gross salary: %s

Personal allowance: %s

Taxable income: %s

%s

Total tax due: %v
`,
		p.Year,
		Amount(p.GrossIncome).String(),
		Amount(p.Allowance).String(),
		Amount(p.TaxableIncome).String(),
		pbstr.String(),
		Amount(p.Due).String(),
	)
}


internal/app/tax-rates.go
package tax

import (
	t "github.com/kerscher/toy-tax-calculator/pkg/tax"
	"github.com/leekchan/accounting"
	"github.com/shopspring/decimal"
)

var (
	DefaultCurrency = accounting.DefaultAccounting("£", 2)

	// MaxIncome is used as a (disregarded) upper bound on last tax
	// bands. The underlying calculation ignores it, but Go has no
	// native concept of nullable types and we give this value so the
	// type checks.
	MaxIncome = t.Amount{
		Value:    decimal.RequireFromString("10000000000000000"),
		Currency: DefaultCurrency,
	}
)

var Rates = map[int]t.Year{
	2019: t.Year{
		Year: 2019,
		Allowance: t.Allowance{
			Value:    decimal.NewFromFloat(11_850),
			Currency: DefaultCurrency,
		},
		Bands: []t.Band{
			t.Band{
				Description: "Starter rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.19)),
				Capacity: t.Amount{
					Value:    decimal.NewFromFloat(2_000),
					Currency: DefaultCurrency,
				},
			},
			t.Band{
				Description: "Basic rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.20)),
				Capacity: t.Amount{
					Value:    decimal.NewFromFloat(10_149),
					Currency: DefaultCurrency,
				},
			},
			t.Band{
				Description: "Intermediate rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.21)),
				Capacity: t.Amount{
					Value:    decimal.NewFromFloat(19_429),
					Currency: DefaultCurrency,
				},
			},
			t.Band{
				Description: "Higher rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.40)),
				Capacity: t.Amount{
					Value:    decimal.NewFromFloat(118_419),
					Currency: DefaultCurrency,
				},
			},
			t.Band{
				Description: "Top rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.46)),
				Capacity:    MaxIncome,
			},
		},
	},
	2018: t.Year{
		Year: 2018,
		Allowance: t.Allowance{
			Value:    decimal.NewFromFloat(11_500),
			Currency: DefaultCurrency,
		},
		Bands: []t.Band{
			t.Band{
				Description: "Basic rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.20)),
				Capacity: t.Amount{
					Value:    decimal.NewFromFloat(31_500),
					Currency: DefaultCurrency,
				},
			},
			t.Band{
				Description: "Higher rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.40)),
				Capacity:    MaxIncome,
			},
		},
	},
	2017: t.Year{
		Year: 2017,
		Allowance: t.Allowance{
			Value:    decimal.NewFromFloat(11_000),
			Currency: DefaultCurrency,
		},
		Bands: []t.Band{
			t.Band{
				Description: "Basic rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.20)),
				Capacity: t.Amount{
					Value:    decimal.NewFromFloat(32_000),
					Currency: DefaultCurrency,
				},
			},
			t.Band{
				Description: "Higher rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.40)),
				Capacity:    MaxIncome,
			},
		},
	},
	2016: t.Year{
		Year: 2017,
		Allowance: t.Allowance{
			Value:    decimal.NewFromFloat(10_600),
			Currency: DefaultCurrency,
		},
		Bands: []t.Band{
			t.Band{
				Description: "Basic rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.20)),
				Capacity: t.Amount{
					Value:    decimal.NewFromFloat(31_785),
					Currency: DefaultCurrency,
				},
			},
			t.Band{
				Description: "Higher rate",
				Rate:        t.Rate(decimal.NewFromFloat(0.40)),
				Capacity:    MaxIncome,
			},
		},
	},
}
